### DISCUSSION TITLE
Can we get a more readable cache.json?

### DISCUSSION THREAD

**Discussion by BettyJJ** [waylaidwanderer/node-chatgpt-api#165] [Discussions]
I didn't change anything in the settings, so the conversations are saved in cache.json as default.

Now the content of the file is like this:
```
{"cache":[["chatgpt:56e4f53f-cc58-491b-8933-03eb66d2ae61",{"value":"{\"value\":{\"messages\":[{\"id\":\"cedd1bdb-6e5f-4595-b34f-e99f27ae8af1\",\"parentMessageId\":\"4b19c269-666b-41c9-a01b-535d3dfd3aa6\",\"role\":\"User\",\"message\":\"Who are you?\"},{\"id\":\"32e9dcf4-733f-48ef-8a9b-8be660053acc\",\"parentMessageId\":\"cedd1bdb-6e5f-4595-b34f-e99f27ae8af1\",\"role\":\"ChatGPT\",\"message\":\"Hello! I'm ChatGPT, a large language model created by OpenAI. Think of me as a virtual assistant that can help answer your questions and provide information on various topics. How can I assist you today?\"}],\"createdAt\":1678099425635},\"expires\":null}"}],["chatgpt:lastConversation",{"value":"{\"value\":{\"conversationId\":\"56e4f53f-cc58-491b-8933-03eb66d2ae61\",\"parentMessageId\":\"32e9dcf4-733f-48ef-8a9b-8be660053acc\"},\"expires\":null}"}]],"lastExpire":1678099411395}
```

It would be a lot more readable if it was like this:
```
{
	"cache": [
		[
			"chatgpt:56e4f53f-cc58-491b-8933-03eb66d2ae61",
			{
				"value": {
					"value": {
						"messages": [
							{
								"id": "cedd1bdb-6e5f-4595-b34f-e99f27ae8af1",
								"parentMessageId": "4b19c269-666b-41c9-a01b-535d3dfd3aa6",
								"role": "User",
								"message": "Who are you?"
							},
							{
								"id": "32e9dcf4-733f-48ef-8a9b-8be660053acc",
								"parentMessageId": "cedd1bdb-6e5f-4595-b34f-e99f27ae8af1",
								"role": "ChatGPT",
								"message": "Hello! I'm ChatGPT, a large language model created by OpenAI. Think of me as a virtual assistant that can help answer your questions and provide information on various topics. How can I assist you today?"
							}
						],
						"createdAt": 1678099425635
					},
					"expires": null
				}
			}
		],
		[
			"chatgpt:lastConversation",
			{
				"value": {
					"value": {
						"conversationId": "56e4f53f-cc58-491b-8933-03eb66d2ae61",
						"parentMessageId": "32e9dcf4-733f-48ef-8a9b-8be660053acc"
					},
					"expires": null
				}
			}
		]
	],
	"lastExpire": 1678099411395
}
```

The main problem is not the newline, but the escape sequences `\"`.  Because of them, I can't use auto formatters to format the file into a readable fashion.

Do you think it is possible to change that? Thank you very much!

**Comment by waylaidwanderer** [comment]
You can code your own storage adapter or use a different third-party one: https://github.com/jaredwray/keyv/blob/main/packages/keyv/README.md#third-party-storage-adapters

Or set your own functions for `keyv-file`'s encode/decode as per the documentation: https://github.com/zaaack/keyv-file

The reason it's formatted like that is because the encode/serialization process turns the data into JSON.

**Reply by BettyJJ** [reply]
Thank you for replying. I'm new to Keyv and Keyv-file. I played around with the code a little, and it seems that the data got JSON.stringify() twice. It's a stringified JSON within a stringified JSON 🤣 Hence the escape sequences `\"`. But I can't pinpoint the exact place in the code where it is done twice.

**Reply by waylaidwanderer** [reply]
I suspect you can try replacing the encode/decode function to simply just return the data instead of encoding it to JSON first.

**Reply by BettyJJ** [reply]
I tried that but it doesn't work. 
I changed line 45 in bin/cli.js to something like this:
```javascript
    const encode = function(v){
        v = JSON.stringify(v, null, '\t');
        return v;
    };
    settings.cacheOptions.store = new KeyvFile({ 
        filename: settings.storageFilePath,
        encode: encode
    });
```
This gives cache.json better indentation.
I can even replace `\"` to `"` but the result will still have an extra pair of quotation marks. 
Returning anything but a string does not work. I think keyv-file requires a string here. So the question is why or where the whole file gets JSON.stringify() again.

**Reply by waylaidwanderer** [reply]
What if you change both encode and decode to simply return the value instead of JSON.stringify/parse?

**Reply by BettyJJ** [reply]
That doesn't work. Only strings are accepted here.
I did some experiments and it seems that it has nothing to do with your code. It's Keyv-file. Keyv-file will just stringify some things twice. 🤣  I changed Keyv-file adding a JSON.parse() somewhere and got what I want.

**Reply by waylaidwanderer** [reply]
Good to know.
